# **Extract, Transform and Load (ETL)**

### Objectives

The purpose for this notebook is to extract, transform and load a raw dataset which will be saved into a clean data frame for Exploratory Data Analysis process. 

1. **Extract** healthcare insurance dataset. Load the raw data into this notebook to begin summarising the DataFrame by applying Pandas methods. 
2. **Transform** check for missing or unique values and duplicated rows to ensure the data is clean prior to further analysis. Standardise text data to ensure consistency for grouping by converting text to lower-case and removing whitespace. Engineering DataFrames adding categorical columns to group together variables.
3. **Load** convert and save clean healthcare insurance DataFrame into a csv file to load in the data visualisation notebook. 

### Input

A CSV healthcare insurance dataset is stored in the Data folder. The dataset is required for the analysis and contains healthcare insurance information that will be used to identify patterns, relationships, and trends within the data.

> File: healthcare_insurance.csv

>File path: /Users/admin/Documents/Code Institute/vscode-projects/Healthcare-Analysis-Project/Healthcare-Analysis-Project/Data/healthcare_insurance.csv

### Import Libraries

Importing dependencies to extract, transform and load data frames for this project including Numpy and Pandas. 

In [1]:
# Importing libraries installed and added to requirements.txt
import numpy as np # Import numpy to perform numerical calculations
import pandas as pd # Import pandas to extract data

## **Extract Data**

### Load DataFrame

Load CSV healthcare insurance dataset using Pandas. I will be referencing the cheatsheet you can access on the link below.

https://pandas.pydata.org/Pandas_Cheat_Sheet.pdf

In [2]:
# Load healthcare_insurance.csv file path 
df = pd.read_csv('/Users/admin/Documents/Code Institute/vscode-projects/Healthcare-Analysis-Project/Healthcare-Analysis-Project/Data/healthcare_insurance.csv')
# Check first 5 rows in the dataframe
df.head()

,age,sex,bmi,children,smoker,region,charges
0,19,female,27.900,0,yes,southwest,16884.92400
1,18,male,33.770,1,no,southeast,1725.55230
2,28,male,33.000,3,no,southeast,4449.46200
3,33,male,22.705,0,no,northwest,21984.47061
4,32,male,28.880,0,no,northwest,3866.85520


### Inspect DataFrame

The code cells below present a summary of the dataframe from different points of view. This intial inspection will help identify attributes which could reduce the quality of the DataFrame.

<br/>

Run <code>df.shape</code> in code cell to check the size of the DataFrame.

In [ ]:
df.shape # Check number of rows and columns 

(1338, 7)


The DataFrame contains 1338 rows, 7 columns. 

<br/>

Run <code>df.shape</code> in code cell to print a summary for the DataFrame. 

In [ ]:
df.info() # Concise summary for the datset 

<class 'pandas.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1338 non-null   int64  
 1   sex       1338 non-null   str    
 2   bmi       1338 non-null   float64
 3   children  1338 non-null   int64  
 4   smoker    1338 non-null   str    
 5   region    1338 non-null   str    
 6   charges   1338 non-null   float64
dtypes: float64(2), int64(2), str(3)
memory usage: 73.3 KB


From this point of view we can summarise the DataFrame as having the folowing attributes: 

* An index ranging between 0 to 1337.

* There are no missing values in each column. 

* Two columns containing decimal numbers (float64), two columns containing whole numbers (int64) and three columns containing text (str). 

* The memory usage is 73.3 KB.

<br/>

Run <code>df.columns</code> to highlight any unsual or incoherent formatting which could be standardised by stripping whitespace and converting text strings into lowercase. 

In [ ]:
df.columns # Check column names

Index(['age', 'sex', 'bmi', 'children', 'smoker', 'region', 'charges'], dtype='str')

The output for <code>df.dtypes</code> has informed me I do not need to format column names. 

<br/>

In [ ]:
df.describe() # Descriptive and statistical overview for each column

,age,bmi,children,charges
count,1338.000000,1338.000000,1338.000000,1338.000000
mean,39.207025,30.663397,1.094918,13270.422265
std,14.049960,6.098187,1.205493,12110.011237
min,18.000000,15.960000,0.000000,1121.873900
25%,27.000000,26.296250,0.000000,4740.287150
50%,39.000000,30.400000,1.000000,9382.033000
75%,51.000000,34.693750,2.000000,16639.912515
max,64.000000,53.130000,5.000000,63770.428010
